files/__addonLocation__/tsconfig.json

```diff
@@ -0,0 +1,20 @@
+{
+  "extends": "@tsconfig/ember/tsconfig.json",
+  "compilerOptions": {
+    "paths": {
+      // No paths, no absolute imports, only node_modules imports allowed
+      // But fallback for type-overrides and such
+      "*": ["types/*"]
+    }
+  },
+  "include": [
+    "src/**/*",
+    "types/**/*"
```

Comment:
In a v1 addon, the root `types` directory is a place we can put non-published type declarations because it actually belongs to the dummy app—it's the exact same `types` directory that we generate for a normal app, except that we messed up and it still ended up in the project root for addons instead of under `tests/dummy`. That oversight is one of the sins I regret most from our early work to support addons in `ember-cli-typescript`, because it creates a blurry picture of the role the `types` directory plays in an addon repository.

As it stands, this change is (re)introducing `types/` as a first-class unpublished directory belonging _only_ to the addon. For the near term, I think we do need some kind of "internal-only type declarations" location for addons—the local Glint registry is a good example of that, though it eventually goes away with strict mode—but my hope would be that the need for such declarations otherwise would be pretty minimal. The more of them you're relying on them, the easier it is to accidentally end up publishing code that won't typecheck for your consumers because they don't have those declarations themselves.

What do you think of something like `local-types/` or `internal-types/`, or even just `internal-types.d.ts` by itself, to help communicate clearly that these types are not part of the addon itself that you're publishing?

Another alternative might be to home `types/` in the test app rather than the addon, but still include it under `includes` so that it's accounted for when typechecking the addon.

/cc @chriskrycho @jamescdavis

Reply:
1. I share the regret. Deeply. :weary:

2. I think we actually have a continuing problem here in that it's actually just really hard to clarify that these types *only* belong to the test infrastructure for the addon and *will not and should not* be published, so I think putting them in the test app is the right move—even something like `local-types` or `internal-types` risks some confusion: "local" or "internal" might still imply to someone that they're not *public* but should be *published* (if that makes sense).

I also wonder (not a blocker for this PR, as we can definitely iterate) whether we might explicitly have something like:

```
/
  app/
    ...
  addon/
    ...
  tsconfig/
    compiler-settings.json
    test.json
    publish.json
tsconfig.json
```

where `tsconfig.json` has `"extends": "./tsconfig/compiler-settings.json"` and the `"includes"` you mention, and then `tsconfig/publish.json` is configured to do the right thing along similar lines. With smart use of comments (thanks, `jsonc!`) in the config files, that combination of naming could help clarify the relationships involved?

It's unfortunate that having those be separate is, if not quite *needful*, then at least potentially helpful, but it's the pattern I've increasingly moved to for handling things like this in my own libraries and I suspect it might be a helpful secondary move here.

Reply:
I see some caveats with regards to moving the internal types to the test app:
* the blueprint creates a single `test-app`, but in some cases people might actually want to have more than one test app, when they need to be different for covering different test cases, e.g. apps with different dependencies (e.g. with and without fastboot). In that case it might become ambiguous and thus confusing where to put these types? (both test-apps might need them)
* things could also go into the other direction, that you don't need a test-app at all. Like when you are able to write plain unit tests for your addon that don't depend on an actual runnable Ember app, you could put those (jest, vitest, whatever) tests right into the addon package, but still need to add some ambient types

Does it maybe make sense to put these type declarations into neither the addon nor the test-app, but basically at `/types` at the root level of the monorepo, outside of any actual package? The downside is that when you need to import some types within your declarations, you would then need to add those to the monorepo's root devDependencies 😕. Or we make that folder a private types-only package that we import from? Just thinking aloud here...

Reply:
I spun up a new addon blueprinted from this branch and played around a bit with possible setups.

After playing with it, I think moving the types directory outside of the addon is a tall order, because for anything registry-based (namely loose-mode Glint, but this would also apply to services or Data), you can wind up running into the same issues that prompted https://github.com/typed-ember/glint/issues/439

The registry pattern almost _requires_ that entries be added somewhere within the package/project that consumes them. Someone who's deeply familiar with both their package manager and TS's resolution scheme may be able to massage their config to get it to work (or if they're on Yarn 1, things may be incidentally hoisted in their favor), but I'd be hesitant to count on that.

Given that, I'm now leaning toward keeping local types within the addon directory for the time being. Maybe we'll figure out an alternative strategy (or maybe one of y'all can come up with one now!), but I wouldn't want to block this indefinitely on that. As far as naming confusion, we could try `unpublished-development-types` or something equally explicit (à la `React.__SECRET_DOM_DO_NOT_USE_OR_YOU_WILL_BE_FIRED`).

I also still kind of like the idea of a single `.d.ts` file rather than a directory, both to discourage too much growth and to give an easy place to leave a `// These types are not part of your published addon, be very careful, blah blah etc` explainer comment. I'm not attached enough to that to fight for it, though 🙂

Reply:
That makes sense to me.

Reply:
For the record, this is now implemented by having a `unpublished-development-types/index.d.ts`, that has a comment like suggested and the boilerplate for the glint registry (for the addon's own classic templates)